### PR TITLE
add prometheus metrics to cache

### PR DESCRIPTION
**Pull request by hendrikKahl**
**What this PR does / why we need it**:

Add capabilities to expose prometheus metrics and an implementation of cache metrics. 

By default, the metrics are not registered to a prometheus registry. For the cache, there is a helper function, which takes a `Registerer` as input. This allows a consumer to decide how to serve metrics.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature developer
Cache exposes usage metrics. Those can be registered to any Prometheus Registry via a helper function.
```
```other dependency
The component-spec has been updated to v0.0.32
```

### CODE REVIEW

File: ociclient/cache/cache.go [hunk 1]
```diff
@@ -46,11 +50,36 @@ func NewCache(log logr.Logger, options ...Option) (Cache, error) {
...
+		if !info.IsDir() {
+			metrics.CachedItems.WithLabelValues(opts.BasePath).Inc()
+			metrics.CacheDiskUsage.WithLabelValues(opts.BasePath).Add(float64(info.Size()))
+		}
+		return nil
+	})
+
+	if err != nil {
+		log.V(7).Info("Error during initialization of cache metrics.")
+		return nil, err
```
Comment (schrodit) [inline ociclient/cache/cache.go:74]: I think we should not log that error as a info.
I would let the caller of the function decide whether he wants to log the error.
```suggestion
		return nil, fmt.Errorf("error during initialization of cache metrics: %w", err)
```

File: ociclient/cache/cache.go [hunk 2]
```diff
@@ -46,11 +50,36 @@ func NewCache(log logr.Logger, options ...Option) (Cache, error) {
 		overlay = memoryfs.New()
 	}
 
+	//initialize metrics
+	metrics.CachedItems.WithLabelValues(opts.BasePath).Set(0)
+	metrics.CacheDiskUsage.WithLabelValues(opts.BasePath).Set(0)
+	metrics.CacheHits.WithLabelValues(opts.BasePath).Add(0)
```
Comment (schrodit) [inline ociclient/cache/cache.go:56]: does it makes sense to have a dedicated hit for the in memory cache?

File: ociclient/cache/cache.go [hunk 3]
```diff
@@ -100,6 +130,14 @@ func (lc *layeredCache) Add(desc ocispecv1.Descriptor, reader io.ReadCloser) err
 	defer file.Close()
 
 	_, err = io.Copy(file, reader)
+
+	// in case everything worked well, update metrics
+	if err == nil {
+		metrics.CachedItems.WithLabelValues(lc.basePath).Inc()
+		if fileInfo, metricsErr := file.Stat(); metricsErr == nil {
```
Comment (schrodit) [inline ociclient/cache/cache.go:137]: I think here we should log the error

File: ociclient/cache/cache.go [hunk 4]
```diff
@@ -167,6 +205,9 @@ func (lc *layeredCache) get(dgst string) (os.FileInfo, vfs.File, error) {
 			lc.log.V(5).Info(err.Error())
 			return info, file, nil
 		}
+		if fileInfo, err := lc.overlayFs.Stat(dgst); err == nil {
```
Comment (schrodit) [inline ociclient/cache/cache.go:208]: lets log the error with a verbosity of 7 or so

**Review by schrodit — CHANGES_REQUESTED** [review]

File: ociclient/cache/cache.go (hunk 2 above)
Comment (hendrikKahl) [reply ociclient/cache/cache.go]: good point, makes totally sense with garbage collection in mind. I've added another vector counter.

**Review by schrodit — APPROVED** [review]
/lgtm

**Pull request merged**